cpp/locality/NeighborList.cc

```diff
@@ -50,8 +51,9 @@ NeighborList::NeighborList(unsigned int num_bonds, const unsigned int* query_poi
         }
         m_neighbors(i, 0) = index;
         m_neighbors(i, 1) = point_index[i];
+        m_vectors[i] = vectors[i];
+        m_distances[i] = std::sqrt(dot(vectors[i], vectors[i]));
```

Comment:
Should we automatically compute this or only when called through `NeighborBond`.

Reply:
This will get fixed when I address the first comment.

Reply:
I would allow this to be done automatically. The neighborlist has to expose its distances to python as a linear array, so it makes sense to store distances and vectors as separate arrays.

I would however, make some changes to the neighborlist class similar to the proposed in NeighborBond to ensure consistency between vectors and distances.

Reply:
I agree that distances in Python need to be a linear array. I was just wondering if for our purposes it would be sufficiently fast to create a NumPy array on the fly when distances are requested. I could even see an argument that distances in Python shouldn't exist as it is a simple NumPy one-liner. Either decision is fine with me though as it isn't generally a costly operation.

Reply:
Creating the arrays on-the-fly seems like more work than is necessary, and honestly may even be a performance hit if a user isn't aware the grabbing the distances is an O(N) operation.

There are non-python places in the code that use the distances, so I'd like to only compute them once and cache them for performance considerations.

Memory usage in the neighborlist isn't a concern either; if it was, this PR would likely not exist.

Reply:
Correct, the design here is precomputing and caching distances because of their common use. I did some very loose calculations/benchmarking and decided this was the right tradeoff (and that the additional memory consumption was irrelevant). Feel free to check me on that, but my recollection is that a C++ cached array of distances is best to precompute because if you get to the point where it is needed, you’re usually in a TBB parallel operation where it would be difficult to compute the whole array AND if you compute the distance at that point, it causes issues with centralizing the floating point correctness. We used to have sqrt’s littered through the various methods’ implementations (when we computed both neighbor vector and distance on the fly) and the conventions and casting were wacky and subtlely different across the code base (not all even used the dot function for the dot product iirc?). That was not good. Having a single source for precomputed distances was better than the previous mess.

Reply:
@bdice @tommy-waltmann I still don't see how this is one source of truth then? `NeighborBonds` also has distances so it is another source of truth. We could for performance reasons say that the vector and distance passed must be compatible and not check, but then we have an untested assumption in `NeighborBond`.

Reply:
There are two places right now where distances are computed: in the `NeighborList` constructor where they are computed from the corresponding vectors and taken directly as an argument for the `NeighborBond` constructor.

I have no problems with computing the distances in the `NeighborList` constructor. It is necessary, in fact, to do this because we need to support custom `NeighborList`s where users give the indices of the neighbors and their vectors. There is no potential for data inconsistency between vectors and distances in the neighborlists because of some recent changes I've made to the neighborlist class on this PR. This design is totally fine with me.

The second place is in the `NeighborBond` class, where both the distance and vector are given directly in the constructor. The design is this way because `NeighborQuery` objects have to compute the distance between a pair of particles to decide if they fit the query arguments before creating the `NeighborBond`. This results in a situation where the distance must be computed before even creating the `NeighborBond`, so to prevent redundant calculations, both the neighbor vector and the distance are given to the `NeighborBond` constructor. In this use case, we trust that the `NeighborQuery` constructs the `NeighborBond` with the correct corresponding distance/vector pair. If other parts of the code were to use this constructor, the opportunity for data inconsistency exists. My plan for this class is to create a constructor for the `NeighborBond` where only the vector is given and the distance is computed automatically, and document that the constructor which takes both the distance and vector should only be used by `NeighborQuery` objects. That seems like the right tradeoff to me in terms of performance vs reducing the potential for errors.

Ultimately, the distances must be able to be computed in multiple places because the `NeighborQuery` and `NeighborList` objects play different roles in iterating over and constructing sets of neighbors at the C++ level. Does that clear things up @b-butler ?

Reply:
I should also add, after having a closer look, that there are also places where `NeighborBond`s are constructed from `NeighborList` entries. This is another case where the correct distance is known beforehand, so re-computing the `NeighborBond` distance would be redundant. I'll heavily document which `NeighborBond` constructor should be used in which case.

Reply:
Yes this makes sense. If we wanted to (not saying it is necessary, but we could use friend functions and a private constructor to ensure we one construct `NeighborBond` objects in methods in select methods.